offchain/eth-input-reader/src/setup.rs

```diff
@@ -1,88 +1,116 @@
 // (c) Cartesi and individual authors (see AUTHORS)
 // SPDX-License-Identifier: Apache-2.0 (see LICENSE)
 
-use eth_state_client_lib::{
-    config::SCConfig, error::StateServerError, BlockServer,
-    GrpcStateFoldClient, StateServer,
-};
+use eth_block_history::{BlockArchive, BlockSubscriber, SubscriptionError};
+use eth_state_fold::StateFoldEnvironment;
 use eth_state_fold_types::BlockStreamItem;
+use ethers::providers::{
+    Http, HttpRateLimitRetryPolicy, Provider, RetryClient,
+};
 use rollups_events::DAppMetadata;
 use snafu::ResultExt;
-use tokio_stream::{Stream, StreamExt};
-use tonic::transport::Channel;
-use types::foldables::authority::{RollupsInitialState, RollupsState};
+use std::sync::{Arc, Mutex};
+use tokio_stream::Stream;
+use types::UserData;
+use url::Url;
 
+use crate::error::{BlockArchiveSnafu, ParseSnafu};
 use crate::{
     config::EthInputReaderConfig,
-    error::{
-        BrokerSnafu, ChannelSnafu, ConnectSnafu, EthInputReaderError,
-        StateServerSnafu,
-    },
+    error::{BrokerSnafu, EthInputReaderError},
     machine::{BrokerStatus, Context},
     metrics::EthInputReaderMetrics,
 };
 
-const BUFFER_LEN: usize = 256;
+/// Maximum events allowed to be in a single provider response. If response
+/// event number reaches over this number, the request must be split into
+/// sub-ranges retried on each of them separately.
+///
+/// Motivation for this configuration parameter mainly comes from Alchemy
+/// as past a certain limit it responds with invalid data.
+pub const MAXIMUM_EVENTS_PER_RESPONSE: usize = 10000;
 
-pub async fn create_state_server(
-    config: &SCConfig,
-) -> Result<
-    impl StateServer<InitialState = RollupsInitialState, State = RollupsState>
-        + BlockServer,
-    EthInputReaderError,
-> {
-    let channel = Channel::from_shared(config.grpc_endpoint.to_owned())
-        .context(ChannelSnafu)?
-        .connect()
-        .await
-        .context(ConnectSnafu)?;
+pub type InputProvider = Provider<RetryClient<Http>>;
+
+pub async fn create_provider(
+    config: &EthInputReaderConfig,
+) -> Result<Arc<InputProvider>, EthInputReaderError> {
+    let http = Http::new(
+        Url::parse(&config.bh_config.http_endpoint).context(ParseSnafu)?,
```

Comment:
Ideally, this parsing should have been inside the `state-fold` crate. Anyway.

I think this is a configuration error (the user provided a bad url). I'd try to "lift" this check to the configuration crate. Also, if you do this, remove the `ParseError` variant from `crate::error::Error`.